cpp/src/parquet/arrow/writer.cc

```diff
@@ -223,30 +221,35 @@ class LevelBuilder {
       const int64_t level_null_count = null_counts_[recursion_level];
       const uint8_t* level_valid_bitmap = valid_bitmaps_[recursion_level];
 
-      for (int64_t i = 0; i < inner_length; i++) {
-        if (i > 0) {
-          RETURN_NOT_OK(rep_levels_.Append(static_cast<int16_t>(rep_level + 1)));
-        }
-        if (level_null_count && level_valid_bitmap == nullptr) {
-          // Special case: this is a null array (all elements are null)
-          RETURN_NOT_OK(def_levels_.Append(static_cast<int16_t>(def_level + 1)));
-        } else if (nullable_level &&
-                   ((level_null_count == 0) ||
-                    BitUtil::GetBit(
-                        level_valid_bitmap,
-                        inner_offset + i + array_offsets_[recursion_level]))) {
-          // Non-null element in a null level
-          RETURN_NOT_OK(def_levels_.Append(static_cast<int16_t>(def_level + 2)));
-        } else {
-          // This can be produced in two case:
-          //  * elements are nullable and this one is null (i.e. max_def_level = def_level
-          //  + 2)
-          //  * elements are non-nullable (i.e. max_def_level = def_level + 1)
-          RETURN_NOT_OK(def_levels_.Append(static_cast<int16_t>(def_level + 1)));
+      if (inner_length >= 1) {
+        RETURN_NOT_OK(
+            rep_levels_.Append(inner_length - 1, static_cast<int16_t>(rep_level + 1)));
+      }
+
+      // Special case: this is a null array (all elements are null)
+      if (level_null_count && level_valid_bitmap == nullptr) {
+        RETURN_NOT_OK(
```

Comment:
Instead of following this block with an `else`, just return here:
```suggestion
        return def_levels_.Append(inner_length, static_cast<int16_t>(def_level + 1));
```

Reply:
Could you also delete the superfluous else on line 218?